lib/rust/ensogl/core/src/gui/component.rs

```diff
@@ -145,11 +145,11 @@ impl<S: Shape> ShapeViewModel<S> {
     }
 
     fn remove_from_scene_layer(&self, old_layer: &WeakLayer) {
-        let flavor = self.data.borrow().flavor();
+        let flavor = (*self.data.borrow()).flavor();
```

Comment:
This change was enough to directly fix the current issue. Everything else is about making it work in more general case (multiple systems of same type on single layer).